features: gssapi principal mapping rules

michael-redpanda opened the pull request:
<!--

See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED.

Describe, in plain language, the motivation behind the change (bug fix,
feature, improvement) in this PR and how the included commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.

  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.

  Backport of PR #PR-NUMBER

-->

Supports parsing and application of Kerberos rules from user config onto Kerberos principals to be used with the ACL.

Changes from force-push `55d45bc`:

* Updated headers
* Added newlines to end of files
* Replaced `const char * const` to `std::string_view`
* Replaced `service_name` with `primary`
* Added `thread_local`
* Directly using `re2` namespace when using `RE2` class calls
* Checks for param size
* Replace `return {}` with `return std::nullopt`

Changes from force-push `4bd9447`:

* Static analysis test failure fix

Changes from force-push `d0b765c`:

* Replaced `make_{upper_lower}` flags with enum

Changes from force-push `18edbc6`:

* Updated log levels/messages

Changes from force-push `0996ea5`:

* Fixed failing `ClusterConfigTest.test_valid_settings` test

Changes from force-push `8b118e5`:

* Fixed python linting error introduced in `0996ea5`

## Backports Required

<!--

Checking at least one of the checkboxes is REQUIRED if this PR is not a backport.

-->

- [x] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

To set Kerberos principal mapping rules:

```
rpk cluster config set sasl_kerberos_principal_mapping '["RULE:[1:$1](App\\..*)s/App\\.(.*)/$1/g","DEFAULT"]'
```

To see the current Kerberos principal mapping rules:

```
rpk cluster config get sasl_kerberos_principal_mapping
```

## Release Notes

<!--

Adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

  ### Bug Fixes

  * Short description of the bug fix if this is a PR to `dev` branch.

  ### Features

  * Short description of the feature. Explain how to configure.

  ### Improvements

  * Short description of how this PR improves existing behavior.

If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

  * none

-->

### Features
* Supports parsing and applying Kerberos rules

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=8257) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=8257) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=8257) it.</sub>

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+
+std::optional<ss::sstring>
+gssapi_rule::apply(std::vector<std::string_view> params) const {
+    static const RE2 non_simple_regex(non_simple_pattern, RE2::Quiet);
```
thread_local?

michael-redpanda replied:
Yeah good catch on that.

Side question -> when the rules are created (when the config binding is updated/set), is that tied to a specific shard?   I guess my question is, is the configuration stuff locked to a specific shard?

BenPope replied:
> > where does this hook in for principal extraction? presumably that is part of the process that occurs after authentication when we map a connection to a user?
> 
> As per what @BenPope said. I started this PR right around when his got merged in. I was under the, possibly wrong, assumption that a separate PR would be required to piece the two together.
> 
> 1. Do you want me to do that in this PR?

I think either way is fine. 

> 2. I still have to add in the correct, I think 'features'(?) into the feature spec.

I don't think anything needs doing here.

> Also, I need to look a little bit more into what the source of "default realm" is. It appears to be a runtime generated item, not set by the user.

I think you'll get the canonical form of the principal as input; it should include the realm.

dotnwat replied:
>Side question -> when the rules are created (when the config binding is updated/set), is that tied to a specific shard? I guess my question is, is the configuration stuff locked to a specific shard?

yes, generally everything is locked to a shard. for this principle mapping stuff i would assume that it happens in the context of a connection, which is itself tied to a specific shard.

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+gssapi_rule::apply(std::vector<std::string_view> params) const {
+    static const RE2 non_simple_regex(non_simple_pattern, RE2::Quiet);
+    const re2::StringPiece match_piece(_match.data(), _match.size());
+    const RE2 match_regex(match_piece, RE2::Quiet);
```
should we use the scoped `re2::RE2`? it's weird that they export RE2 into the global scope in the headers...

michael-redpanda replied:
I can do that for sure.  I agree it is weird, and not a fan of 3rd party libraries bringing things into scope outside of my control.

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+    ss::sstring result;
+
+    if (_is_default) {
+        if (_default_realm == params[0]) {
```
checked access (e.g. `params.at(0)` since there doesn't seem to be any handling for params being equal?

michael-redpanda replied:
Sorry did you mean "empty" instead of "equal"?  If so you're correct and will add that check.

dotnwat replied:
yes, i meant empty

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    if (_to_lower && !result.empty()) {
+        boost::algorithm::to_lower(result, std::locale::classic());
+    }
+
+    if (_to_upper && !result.empty()) {
+        boost::algorithm::to_upper(result, std::locale::classic());
```
is it true that to_lower and to_upper are mutually exclusive values? if so, maybe use a single bool_class?

michael-redpanda replied:
Yeah, I admit I copied this style from another piece of code.  Using an enum is definitely 'more correct'.

dotnwat replied:
if they aren't mutually exclusive, then an enum makes sense. if they are mutually exclusive then it seems like a single bool is sufficient.

BenPope replied:
It's tri-state; upper, lower, ore leave alone.

dotnwat replied:
ahh, optional<bool> or enum then?

michael-redpanda replied:
`enum case_change_operation { noop, make_lower, make_upper };`

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+                  seclog.error,
+                  "GSSAPI: Invalid number provided for username format: {}",
+                  std::string_view{index_replace.data(), index_replace.size()});
+                return {};
```
return std::nullopt ?

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+            }
+        }
+
+        if (format_piece.length() <= 0) {
```
format_piece.empty() ?

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+            result = ss::sstring(params[1]);
+        }
+    } else if (
+      params.size() - 1 == static_cast<size_t>(_number_of_components)) {
```
there probably needs to be some checking some where that params conforms to correct input. some branches it needs to be len()>=2 some needs to be non-empty, etc...

dotnwat commented on src/v/security/tests/gssapi_principal_mapper_test.cc:
```diff
@@ -0,0 +1,67 @@
+    }
+};
+
+static std::array<gssapi_test_record, 3> gssapi_name_test_data{
```
you can use `std::to_array` to avoid having to encode the size `3`

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -283,6 +359,42 @@ ss::sstring gssapi_rule::replace_substitution(
+        }
+    }
+
+    vlog(seclog.error, "No rules apply to {}, rules: {}", name, _rules);
```
is this really an error? the message doesn't sound like an error, but i'm also not certain of what the scenario is.

michael-redpanda replied:
That's fair.  I feel like it's "more important" than an informational log so maybe warning level?

dotnwat replied:
maybe. i sent you a link to log level guide on slack

michael-redpanda replied:
Pushed updates and commented (in the review) the `vlogs` with my justification for the levels.  Happy to hear your take on it.

dotnwat commented on src/v/security/gssapi_principal_mapper.h:
```diff
@@ -108,6 +109,26 @@ class gssapi_rule {
+private:
+    friend struct fmt::formatter<gssapi_principal_mapper>;
+
+    friend std::ostream&
+    operator<<(std::ostream& os, const gssapi_principal_mapper& p);
```
i'm curious: if a fmt::formatter specialization is defined, is the output operator needed?

michael-redpanda replied:
No it is not.  Just commented it out and built/ran unit tests and they ran fine.  I'll remove them.

michael-redpanda replied:
Actually nix that.  I removed the one for `gssapi_rule` and it didn't build.  I need to look a little bit more into how fmt works with this.  My only experience with fmt is using it with `spdlog`

dotnwat replied:
i wouldn't waste much time on it if it is working. at this point it's a curiosity if we can have one or the other in the future.

dotnwat reviewed (commented):
looks nice!

michael-redpanda commented:
I think I forgot one important thing: fully tying this into the configuration system.  Will do that tomorrow.

dotnwat commented:
> I think I forgot one important thing: fully tying this into the configuration system. Will do that tomorrow.

where does this hook in for principal extraction? presumably that is part of the process that occurs after authentication when we map a connection to a user?

michael-redpanda commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+  , _number_of_components(number_of_components)
+  , _format(format)
+  , _match(match)
+  , _from_pattern(std::regex{
+      from_pattern.begin(),
+      from_pattern.length(),
```
`from_pattern` is a string view and I don't believe you can go directly from `string_view` to `basic_string` without performing a copy.  I could pass it in like `from_pattern.data(), from_pattern.size()` but I prefer the iterator type constructor (see https://en.cppreference.com/w/cpp/regex/basic_regex/basic_regex)

dotnwat replied:
huh, kinda annoying that std::regex doesn't have a string_view constructor. oh well!

BenPope commented:
> > I think I forgot one important thing: fully tying this into the configuration system. Will do that tomorrow.
> 
> where does this hook in for principal extraction? presumably that is part of the process that occurs after authentication when we map a connection to a user?

Around here: https://github.com/redpanda-data/redpanda/blob/1cced14802d9b9521050d04841c671715e12d2b6/src/v/security/gssapi_authenticator.cc#L402

BenPope commented on src/v/security/gssapi_principal_mapper.h:
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
```
This should also be the RCL.

michael-redpanda commented:
> where does this hook in for principal extraction? presumably that is part of the process that occurs after authentication when we map a connection to a user?

As per what @BenPope said.  I started this PR right around when his got merged in.  I was under the, possibly wrong, assumption that a separate PR would be required to piece the two together.

1. Do you want me to do that in this PR?
2. I still have to add in the correct, I think 'features'(?) into the feature spec.

Also, I need to look a little bit more into what the source of "default realm" is.  It appears to be a runtime generated item, not set by the user.

BenPope commented on src/v/security/tests/gssapi_principal_mapper_test.cc:
```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
```
May as well make this RCL, too.

BenPope commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,474 @@
+} // namespace detail
+
+gssapi_name::gssapi_name(
+  std::string_view service_name,
```
nit: This might better be called `primary` as I imagine it could be a user or service.

michael-redpanda replied:
I was matching other source that was calling this "service_name" but I can replace it with primary if you prefer.

michael-redpanda commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
```
Understood.  I grabbed this header from some other source file.  Do all of those headers need to be replaced as well?

michael-redpanda commented on src/v/security/gssapi_principal_mapper.h:
```diff
@@ -0,0 +1,100 @@
+    template<typename FormatContext>
+    typename FormatContext::iterator
+    format(const type& r, FormatContext& ctx) const;
+};
```
Probably a CLion thing.  Let me double check some settings.

michael-redpanda commented:
> > 2\. I still have to add in the correct, I think 'features'(?) into the feature spec.
>
> I don't think anything needs doing here

I need to check but I think `sasl_kerberos_principal_mapping` needs to be added?

> > Also, I need to look a little bit more into what the source of "default realm" is. It appears to be a runtime generated item, not set by the user.
> 
> I think you'll get the canonical form of the principal as input; it should include the realm.

`default_realm` is actually something different than the realm provided by the kerberos principal (which may or may not provide a realm).

NyaliaLui commented on src/v/security/tests/gssapi_principal_mapper_test.cc:
```diff
@@ -0,0 +1,67 @@
+      BOOST_REQUIRE_EQUAL(c.expected_realm, name.realm());
+      BOOST_REQUIRE_EQUAL(c.gssapi_principal_name, fmt::format("{}", name)););
+}
+} // namespace security
```
nit: make sure there is an empty line at the end of c/c++ files

NyaliaLui commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -341,4 +458,17 @@ fmt::formatter<security::gssapi_rule, char, void>::format<
+      "[{}] (default_realm: {})",
+      fmt::join(r._rules, ", "),
+      r._default_realm);
 }
```
nit: newline

michael-redpanda commented:
Force push:

* Updated headers
* Added newlines to end of files
* Replaced `const char * const` to `std::strig_view`
* Replaced `service_name` with `primary`
* Added `thread_local`
* Directly using `re2` namespace when using `RE2` class calls
* Checks for param size
* Replace `return {}` with `return std::nullopt`

michael-redpanda commented:
Force pushed static analysis test failure fix

dotnwat commented:
>As per what @BenPope said. I started this PR right around when his got merged in. I was under the, possibly wrong, assumption that a separate PR would be required to piece the two together.
> Do you want me to do that in this PR?

I think it is up to you. Since the code in this PR has a test, but is otherwise not hooked up to any of the rest of the redpanda (that's correct, right?) then there is no risk in merging and the following up with adding the connective tissue.

michael-redpanda commented:
> I think it is up to you. Since the code in this PR has a test, but is otherwise not hooked up to any of the rest of the redpanda (that's correct, right?) then there is no risk in merging and the following up with adding the connective tissue.

I plan on pushing one more change that allows user configuration of the kerberos/gssapi rules via rpk (effectively updating configuration.cc).  Just need to write an e2e test for it.

emaxerrno commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,482 @@
+    }
+
+    std::vector<gssapi_rule> rv;
+    re2::StringPiece default_;
```
random thought, i wonder if stringpiece is compatible w/ string_view

dotnwat replied:
our internal version of re2 is, but not the one that ships in the latest fedora.

michael-redpanda replied:
Yeah this was the source of my confusion/heartburn earlier today.

emaxerrno commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,482 @@
+              &to_pattern,
+              &repeat,
+              &upper_lower)) {
+            throw std::runtime_error("GSSAPI: Invalid rule: " + rule);
```
this needs a prometheus metric.

dotnwat replied:
@michael-redpanda i think the metrics can be collected at the kafka layer as these errors bubble up to the connection level?

michael-redpanda replied:
This logic verifies that the rule provided by the _user_ is good and is not when the rule is actually applied to the kerberos principal.

As in, checks that "RULE:[1:$1](redpanda.*)s/(redpanda).*/$1rocks/g" is good.

michael-redpanda replied:
> @michael-redpanda i think the metrics can be collected at the kafka layer as these errors bubble up to the connection level?

I don't know :)  Would love to have a conversation about this to better understand what you're looking for

michael-redpanda commented:
Latest commit to allow configuration of Kerberos principal mapping.  I still need to write e2e tests for this but having issues getting that running on my box but didn't want to hold up the PR.

michael-redpanda commented:
Force-push: Updated log levels/messages

michael-redpanda commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,505 @@
+    if (
+      !result.empty()
+      && re2::RE2::PartialMatch(result_piece, non_simple_regex)) {
+        vlog(
+          seclog.error,
+          "Non-simple name {} after auth_to_local rule {}",
+          result,
+          *this);
```
Logged as an error because this indicates an issue with the rule itself.  The rule should never contain an `@`

michael-redpanda commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,505 @@
+            auto conv_result = std::from_chars(
+              index_replace.begin(), index_replace.end(), index);
+            if (conv_result.ec != std::errc()) {
+                vlog(
+                  seclog.warn,
+                  "Bad format in username mapping in {}",
+                  std::string_view{index_replace.data(), index_replace.size()});
+                return std::nullopt;
+            } else if (index >= params.size()) {
+                vlog(
+                  seclog.warn,
+                  "Invalid index provided ({}).  Outside range of "
+                  "parameters (length {})",
+                  index,
+                  params.size());
```
Warning -> User provided a bad rule to us but is 'technically' valid.

dotnwat replied:
i don't have a good intuition about log levels in general, but im wondering why warning--it doesn't seem like redpanda itself is behaving in a concerning way, right?

michael-redpanda commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,505 @@
+        }
+    }
+
+    vlog(seclog.warn, "No rules apply to {}, rules: {}", name, _rules);
```
A kerberos principal used to authenticate did not match any rule -> User should see this and update rules

dotnwat replied:
The downside to logging like this is that it lacks context. For example, what connection did this occur on? One way to solve that is to return an appropriate error or throw an exception. Then in the kafka connection context we can log in a way that annotates the log message with connection context information.

If a principal doesn't match a rule, then having that connection information could be very useful: a system with 1000s of connection might be hard to figure out where the rogue principal is coming from. With connection information we can provide things like the kafka client ID and IP address of the offending user.

No need to change this right now in this PR, but something to consider in follow-ups.

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -0,0 +1,286 @@
+            }
+        }
+
+        // format_piece will be empty if we have parsed all of format_piece
+        // text_replace and index_replace will be empty if nothing was grouped
+        // into those but re2::RE2::Consume still returns true (and doesn't
+        // update format_piece)
```
great comment. all those in/out parameters with consume from re2 and std::from_chars... not easy to follow!

dotnwat commented on src/v/security/gssapi_principal_mapper.h:
```diff
@@ -0,0 +1,96 @@
+
+class gssapi_rule {
+public:
+    enum case_change_operation { noop, make_lower, make_upper };
```
prefer `enum class` over `enum`. the former is a strong type.

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -21,9 +21,57 @@
+          nullptr,
+          &host_name,
+          &realm)) {
+        return gssapi_name(
+          ss::sstring(primary), ss::sstring(host_name), ss::sstring(realm));
```
nit: you could return `{ ... }` here instead of repeating the name of the function's return type

dotnwat commented on src/v/security/gssapi_principal_mapper.cc:
```diff
@@ -292,6 +376,42 @@ ss::sstring gssapi_rule::replace_substitution(
+  , _principal_to_local_rules_binding(std::move(principal_to_local_rules_cb))
+  , _rules{detail::parse_rules(
+      _default_realm, _principal_to_local_rules_binding())} {
+    _principal_to_local_rules_binding.watch([this]() {
+        _rules = detail::parse_rules(
+          _default_realm, _principal_to_local_rules_binding());
```
+1
